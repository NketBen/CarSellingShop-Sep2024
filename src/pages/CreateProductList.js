//componente que sube todos items a firestore 

import { useState } from "react";
import Spinner from "../components/Spinner";
import { toast } from "react-toastify";
import {getStorage, ref, uploadBytesResumable, getDownloadURL} from "firebase/storage";
import { getAuth } from "firebase/auth";
import { v4 as uuidv4 } from "uuid";
import { addDoc, collection, serverTimestamp } from "firebase/firestore";
import { db } from "../firebase";
import { useNavigate } from "react-router-dom";

//componente para crear items
export default function CreateProductList() { 
  const navigate = useNavigate();
  const auth = getAuth();
  const [geolocationEnabled, setGeolocationEnabled] = useState(true);
  const [loading, setLoading] = useState(false);
  const [paraForm, SetParaForm] = useState({
    type: "new",
    name: "",
    seats: 1,
    previousOwners: 1,
    MOT: false,
    defect: false,
    address: "",
    description: "",
    offer: false,
    realPrice: 0,
    discountedPrice: 0,
    latitude: 0,
    longitude: 0,
    images: {},
  });
  const {
    type,
    name,
   seats,
    previousOwners,
    MOT,
    address,
    defect,
    description,
    offer,
    realPrice,
    discountedPrice,
    latitude,
    longitude,
    images,
  } = paraForm;
  function onChange(e) {
    let boolean = null;
    if (e.target.value === "true") {
      boolean = true;
    }
    if (e.target.value === "false") {
      boolean = false;
    }
    // Files
    if (e.target.files) {
      SetParaForm((prevState) => ({
        ...prevState,
        images: e.target.files,
      }));
    }
    // Text/Booleano/Number
    if (!e.target.files) {
      SetParaForm((prevState) => ({
        ...prevState,
        [e.target.id]: boolean ?? e.target.value,
      }));
    }
  }
  async function onSubmit(e) {
    e.preventDefault();
    setLoading(true);
    if (+discountedPrice >= +realPrice) { //si el precio de rebajada es major que precio de choche
      setLoading(false);
      toast.error("Precio de Descuento debe ser mas que precio de Oferta");
      return;
    }
    if (images.length > 10) { // solo aceptamos hasta 10 imagenes
      setLoading(false);
      toast.error("No se permite mas de 10 imagenes");
      return;
    }
    let geolocation = {};
    let location;
    if (geolocationEnabled) {
      //obtener coordinada de direccion en la formulario
      const response = await fetch( 
        `https://maps.googleapis.com/maps/api/geocode/json?address=${address}&key=${process.env.REACT_APP_GEOCODE_API_KEY}` //google map api
      );
      const data = await response.json();
      geolocation.lat = data.results[0]?.geometry.location.lat ?? 0;
      geolocation.lng = data.results[0]?.geometry.location.lng ?? 0;

      location = data.status === "ZERO_RESULTS" && undefined;

      if (location === undefined) {
        setLoading(false);
        toast.error("Por favor, revisa tu dirrecion");
        return;
      }
    } else {
      geolocation.lat = latitude;
      geolocation.lng = longitude;
    }

    //creamos una funcion para subir imagenes
    async function storeImage(image) {
      return new Promise((resolve, reject) => {
        const storage = getStorage();
        const filename = `${auth.currentUser.uid}-${image.name}-${uuidv4()}`;
        const storageRef = ref(storage, filename); // referencia de storage
        const uploadTask = uploadBytesResumable(storageRef, image);
        uploadTask.on(
          "state_changed",
          (snapshot) => {
            const progress =
              (snapshot.bytesTransferred / snapshot.totalBytes) * 100;
            switch (snapshot.state) {
              case "paused":
                break;
              case "running":
                break;
                default:console.log("Something is wrong");
            }
          },
          (error) => {
            reject(error);
          },
          () => {
            getDownloadURL(uploadTask.snapshot.ref).then((downloadURL) => {
              resolve(downloadURL);
            });
          }
        );
      });
    }
    //tener url de imagenes subida
    const imgUrls = await Promise.all(
      [...images].map((image) => storeImage(image))
    ).catch((error) => {
      setLoading(false);
      toast.error("Imagenes subido");
      return;
    });

    const paraFormCopy = {
      ...paraForm,
      imgUrls,
      geolocation,
      timestamp: serverTimestamp(),
      userRef: auth.currentUser.uid,
    };
    delete paraFormCopy.images;
    !paraFormCopy.offer && delete paraFormCopy.discountedPrice;
    delete paraFormCopy.latitude;
    delete paraFormCopy.longitude;
    const docRef = await addDoc(collection(db, "items"), paraFormCopy);// subir imagenes url y todos a firestore
    setLoading(false);
    toast.success("Se ha creado listado de item correctamente");
    navigate(`/category/${paraFormCopy.type}/${docRef.id}`) //navegar a categoria de item subida
   
  }

  if (loading) {
    return <Spinner />;
  }
  return (
    <main className="max-w-md px-2 mx-auto">
      <h1 className="text-3xl text-center mt-6 font-bold">Crear tu Listas de ventas</h1>
      <form onSubmit={onSubmit}>
        <p className="text-lg mt-6 font-semibold">secondHand / new</p>
        <div className="flex">
          <button
            type="button"
            id="type"
            value="sale"
            onClick={onChange}
            className={`mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              type === "new"
                ? "bg-white text-black"
                : "bg-slate-600 text-white"
            }`}
          >
            secondHand
          </button>
          <button
            type="button"
            id="type"
            value="new"
            onClick={onChange}
            className={`ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              type === "sale"
                ? "bg-white text-black"
                : "bg-slate-600 text-white"
            }`}
          >
            New
          </button>
        </div>
        <p className="text-lg mt-6 font-semibold">Nombre De Coche</p>
        <input
          type="text"
          id="name"
          value={name}
          onChange={onChange}
          placeholder="Nombre de Coche"
          maxLength="32"
          minLength="5"
          required
          className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 mb-6"
        />
        <div className="flex space-x-6 mb-6">
          <div>
            <p className="text-lg font-semibold">Seats</p>
            <input
              type="number"
              id="seats"
              value={seats}
              onChange={onChange}
              min="1"
              max="50"
              required
              className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
            />
          </div>
         { type!== "new" && (
             <div>
            <p className="text-lg font-semibold">Due√±os Anteriores</p>
            <input
              type="number"
              id="previousOwners"
              value={previousOwners}
              onChange={onChange}
              min="1"
              max="50"
              required
              className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
            />
          </div>
          )}
        </div>
        <p className="text-lg mt-6 font-semibold">MOT Pass</p>
        <div className="flex">
          <button
            type="button"
            id="MOT"
            value={true}
            onClick={onChange}
            className={`mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              !MOT? "bg-white text-black" : "bg-slate-600 text-white"
            }`}
          >
            Yes
          </button>
          <button
            type="button"
            id="MOT"
            value={false}
            onClick={onChange}
            className={`ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              MOT? "bg-white text-black" : "bg-slate-600 text-white"
            }`}
          >
            No
          </button>
        </div>
        <p className="text-lg mt-6 font-semibold">Defect</p>
        <div className="flex">
          <button
            type="button"
            id="defect"
            value={true}
            onClick={onChange}
            className={`mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              !defect ? "bg-white text-black" : "bg-slate-600 text-white"
            }`}
          >
          Yes
          </button>
          <button
            type="button"
            id="defect"
            value={false}
            onClick={onChange}
            className={`ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              defect ? "bg-white text-black" : "bg-slate-600 text-white"
            }`}
          >
            No
          </button>
        </div>
        <p className="text-lg mt-6 font-semibold">Address</p>
        <textarea
          type="text"
          id="address"
          value={address}
          onChange={onChange}
          placeholder="Direccion"
          required
          className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 mb-6"
        />
        {!geolocationEnabled && (
          <div className="flex space-x-6 justify-start mb-6">
            <div className="">
              <p className="text-lg font-semibold">Latitude</p>
              <input
                type="number"
                id="latitude"
                value={latitude}
                onChange={onChange}
                required
                min="-90.0000"
                max="90.0000"
                className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:bg-white focus:text-gray-700 focus:border-slate-600 text-center"
              />
            </div>
            <div className="">
              <p className="text-lg font-semibold">Longitude</p>
              <input
                type="number"
                id="longitude"
                value={longitude}
                onChange={onChange}
                required
                min="-180.0000"
                max="180.000"
                className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:bg-white focus:text-gray-700 focus:border-slate-600 text-center"
              />
            </div>
          </div>
        )}
        <p className="text-lg font-semibold">Descripci√≥n</p>
        <textarea
          type="text"
          id="description"
          value={description}
          onChange={onChange}
          placeholder="Descripci√≥n"
          required
          className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 mb-6"
        />
        <p className="text-lg font-semibold">Oferta</p>
        <div className="flex mb-6">
          <button
            type="button"
            id="offer"
            value={true}
            onClick={onChange}
            className={`mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              !offer ? "bg-white text-black" : "bg-slate-600 text-white"
            }`}
          >
           Si
          </button>
          <button
            type="button"
            id="offer"
            value={false}
            onClick={onChange}
            className={`ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full ${
              offer ? "bg-white text-black" : "bg-slate-600 text-white"
            }`}
          >
            No
          </button>
        </div>
        <div className="flex items-center mb-6">
          <div className="">
            <p className="text-lg font-semibold">Precio Real</p>
            <div className="flex w-full justify-center items-center space-x-6">
              <input
                type="number"
                id="realPrice"
                value={realPrice}
                onChange={onChange}
                min="100"
                max="600000000"
                required
                className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
              />
              {type === "new" && (
                <div className="">
                  <p className="text-md w-full whitespace-nowrap">‚Ç¨</p>
                </div>
              )}
            </div>
          </div>
        </div>
        {offer && (
          <div className="flex items-center mb-6">
            <div className="">
              <p className="text-lg font-semibold">Precio Ofertado</p>
              <div className="flex w-full justify-center items-center space-x-6">
                <input
                  type="number"
                  id="discountedPrice"
                  value={discountedPrice}
                  onChange={onChange}
                  min="50"
                  max="400000000"
                  required={offer}
                  className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
                />
                {type === "new" && (
                  <div className="">
                    <p className="text-md w-full whitespace-nowrap">
                     ‚Ç¨
                    </p>
                  </div>
                )}
              </div>
            </div>
          </div>
        )}
        <div className="mb-6">
          <p className="text-lg font-semibold">Images</p>
          <p className="text-gray-600">
            Primer imagen sera la cubierta (max 10)
          </p>
          <input
            type="file"
            id="images"
            onChange={onChange}
            accept=".jpg,.png,.jpeg"
            multiple
            required
            className="w-full px-3 py-1.5 text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:bg-white focus:border-slate-600"
          />
        </div>
        <button
          type="submit"
          className="mb-6 w-full px-7 py-3 bg-blue-600 text-white font-medium text-sm uppercase rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
        >
          Crear tu Listas
        </button>
      </form>
    </main>
  );
}
