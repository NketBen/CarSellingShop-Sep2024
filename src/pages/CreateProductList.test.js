


import React from "react";
import {render, screen} from "@testing-library/react";
import CreateProductList from "./CreateProductList";
import '@testing-library/jest-dom'
import { BrowserRouter } from "react-router-dom";
import renderer from 'react-test-renderer'

//hacemos test para comprobar que el elemento h1  con texto "Crear tu Listas de ventas" existe 
it('render a message', ()=> {
    render(<BrowserRouter><CreateProductList message="Crear tu Listas de ventas"/></BrowserRouter>);
    const messageElement=screen.getByText(/Crear tu Listas de ventas/i);
    expect(messageElement).toBeInTheDocument();

});

//hacemos test para comprobar que el componente  CreateProductList.js no tiene comportamiento inesperada
it('render correctly',()=>{
    const component=renderer.create(<BrowserRouter><CreateProductList/></BrowserRouter>);
   const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});
