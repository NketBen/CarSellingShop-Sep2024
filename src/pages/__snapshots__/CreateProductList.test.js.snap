// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<main
  className="max-w-md px-2 mx-auto"
>
  <h1
    className="text-3xl text-center mt-6 font-bold"
  >
    Crear tu Listas de ventas
  </h1>
  <form
    onSubmit={[Function]}
  >
    <p
      className="text-lg mt-6 font-semibold"
    >
      secondHand / new
    </p>
    <div
      className="flex"
    >
      <button
        className="mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-white text-black"
        id="type"
        onClick={[Function]}
        type="button"
        value="sale"
      >
        secondHand
      </button>
      <button
        className="ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-slate-600 text-white"
        id="type"
        onClick={[Function]}
        type="button"
        value="new"
      >
        New
      </button>
    </div>
    <p
      className="text-lg mt-6 font-semibold"
    >
      Nombre De Coche
    </p>
    <input
      className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 mb-6"
      id="name"
      maxLength="32"
      minLength="5"
      onChange={[Function]}
      placeholder="Nombre de Coche"
      required={true}
      type="text"
      value=""
    />
    <div
      className="flex space-x-6 mb-6"
    >
      <div>
        <p
          className="text-lg font-semibold"
        >
          Seats
        </p>
        <input
          className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
          id="seats"
          max="50"
          min="1"
          onChange={[Function]}
          required={true}
          type="number"
          value={1}
        />
      </div>
    </div>
    <p
      className="text-lg mt-6 font-semibold"
    >
      MOT Pass
    </p>
    <div
      className="flex"
    >
      <button
        className="mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-white text-black"
        id="MOT"
        onClick={[Function]}
        type="button"
        value={true}
      >
        Yes
      </button>
      <button
        className="ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-slate-600 text-white"
        id="MOT"
        onClick={[Function]}
        type="button"
        value={false}
      >
        No
      </button>
    </div>
    <p
      className="text-lg mt-6 font-semibold"
    >
      Defect
    </p>
    <div
      className="flex"
    >
      <button
        className="mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-white text-black"
        id="defect"
        onClick={[Function]}
        type="button"
        value={true}
      >
        Yes
      </button>
      <button
        className="ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-slate-600 text-white"
        id="defect"
        onClick={[Function]}
        type="button"
        value={false}
      >
        No
      </button>
    </div>
    <p
      className="text-lg mt-6 font-semibold"
    >
      Address
    </p>
    <textarea
      className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 mb-6"
      id="address"
      onChange={[Function]}
      placeholder="Direccion"
      required={true}
      type="text"
      value=""
    />
    <p
      className="text-lg font-semibold"
    >
      Descripción
    </p>
    <textarea
      className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 mb-6"
      id="description"
      onChange={[Function]}
      placeholder="Descripción"
      required={true}
      type="text"
      value=""
    />
    <p
      className="text-lg font-semibold"
    >
      Oferta
    </p>
    <div
      className="flex mb-6"
    >
      <button
        className="mr-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-white text-black"
        id="offer"
        onClick={[Function]}
        type="button"
        value={true}
      >
        Si
      </button>
      <button
        className="ml-3 px-7 py-3 font-medium text-sm uppercase shadow-md rounded hover:shadow-lg focus:shadow-lg active:shadow-lg transition duration-150 ease-in-out w-full bg-slate-600 text-white"
        id="offer"
        onClick={[Function]}
        type="button"
        value={false}
      >
        No
      </button>
    </div>
    <div
      className="flex items-center mb-6"
    >
      <div
        className=""
      >
        <p
          className="text-lg font-semibold"
        >
          Precio Real
        </p>
        <div
          className="flex w-full justify-center items-center space-x-6"
        >
          <input
            className="w-full px-4 py-2 text-xl text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:text-gray-700 focus:bg-white focus:border-slate-600 text-center"
            id="realPrice"
            max="600000000"
            min="100"
            onChange={[Function]}
            required={true}
            type="number"
            value={0}
          />
          <div
            className=""
          >
            <p
              className="text-md w-full whitespace-nowrap"
            >
              €
            </p>
          </div>
        </div>
      </div>
    </div>
    <div
      className="mb-6"
    >
      <p
        className="text-lg font-semibold"
      >
        Images
      </p>
      <p
        className="text-gray-600"
      >
        Primer imagen sera la cubierta (max 10)
      </p>
      <input
        accept=".jpg,.png,.jpeg"
        className="w-full px-3 py-1.5 text-gray-700 bg-white border border-gray-300 rounded transition duration-150 ease-in-out focus:bg-white focus:border-slate-600"
        id="images"
        multiple={true}
        onChange={[Function]}
        required={true}
        type="file"
      />
    </div>
    <button
      className="mb-6 w-full px-7 py-3 bg-blue-600 text-white font-medium text-sm uppercase rounded shadow-md hover:bg-blue-700 hover:shadow-lg focus:bg-blue-700 focus:shadow-lg active:bg-blue-800 active:shadow-lg transition duration-150 ease-in-out"
      type="submit"
    >
      Crear tu Listas
    </button>
  </form>
</main>
`;
